examples/amm/lib.rs

```diff
@@ -0,0 +1,373 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+const PRECISION: u128 = 1_000_000;
+
+#[ink::contract]
+mod amm {
+    use ink_storage::{traits::SpreadAllocate, Mapping};
+
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        ZeroLiquidity,
+        ZeroAmount,
+        InsufficientAmount,
+        NonEquivalentAmount,
+        ThresholdNotReached,
+        InvalidShare,
+        InsufficientLiquidity,
+        SlippageExceeded,
+    }
+
+    #[derive(Default, SpreadAllocate)]
+    #[ink(storage)]
+    pub struct Amm {
+        total_shares: Balance,
+        total_token1: Balance,
+        total_token2: Balance,
+        token1_balance: Mapping<AccountId, Balance>,
+        token2_balance: Mapping<AccountId, Balance>,
+        lp_shares: Mapping<AccountId, Balance>,
+        fee: Balance,
+    }
+
+    #[ink(impl)]
+    impl Amm {
+        fn has_valid_amount(
+            &self,
+            balance: &Mapping<AccountId, Balance>,
+            _qty: Balance,
```

Comment:
We don't use the same underscore convention as in Solidity - the Rust compiler will make it clear if you're using the variable in the wrong context. Rust also uses [underscores in a different context](https://stackoverflow.com/questions/48361537/why-do-underscore-prefixed-variables-exist#48361729), which makes it confusing for not only the compiler but also other Rust developers